Implement Merkle-Tree Level hashing & kelvin traits inclusion

CPerezz opened the pull request:
- [x] Add merkle tree hashing logic for `Scalar` and `Variable` as a gadget. Closes #9 
- [x] Implement needed Kelvin-related Traits. Closes #12
- [x] Implement `From<Branch> for `PoseidonBranch`. Close #13
- [x] Create a wrapping type over Scalar to define the `Content` trait over it. Closes #14

Commit:
Implement Merkle Tree-level Hashing methods

Implements the Poseidon Hash for a Merkle Tree
level for both `Scalar` and `Gadget` strategies.

This relies on `perm` function exported from `Hades252`.
It includes the logic required by the Poseidon Hash
which computes the bitflags in respect of the presence of
the level leafs.

Actually, this functions support the merkle-level hashing
for `Scalar` & `Variable`.

The maximum ammount of leafs per level "ARITY" is actually
4. So for bigger trees, this functions will not behave
correctly. This is a tradeoff we need to do until we can
support custom WIDTH permutations in Hades252 repo.

Added a test to check that the implementations of
merkle_level_gadget & merkle_level_hash_gadget_without_bitflags
inside of a PLONK proof.

The test check that the implementation of the proof is consistent
against the Scalar merkle_level_hash implementation.

Commit:
Implement PoseidonBranch & PoseidonLevel

The `PoseidonBranch` is the data structure over which
we will build our opening proofs.

- Implement `From<kelvin::Branch> for PoseidonBranch.
- Implement `PoseidonLevel` structure.
- Add docs for every structure.

These two allow us to obtain in the correct structure
the data obtained from a `kelvin` search to then elaborate
a merkle-opening proof.

We needed a way to collect the data from the kelvin `Branch` and
store it on a way which allows an easy Proof generation on later
stages.

The `From` trait implementation takes care of iterating over the Branch
levels and so, collect all of the data while also computes the bitflags
terms for each one of the leves. It also stores the index of the
position of the upper level hash to link the results and chain up the
hashing procedure until the root of the tree.

- Implemented also a few extra methods that allow to generate default
empty poseidon branches and levels so we have easier ways to generate
the `PoseidonBranch` in the `From` impl.

- Move ARITY definition to `lib.rs`.
- Add some module-docs.

Commit:
Add hashing_utils mod & define `ScalarStorage`

In order to allow Kelvin understand how to store `Scalar`
in disk, we needed to implement a Wrapper type over `Scalar`
which implements `kelvin::Content`.

- Define `StorageScalar` which is a simple wrapper over `Scalar` type
of `dusk-bls12_381`.

- Implemented some additional helpers to convert between `Scalar`
& `StorageScalar`

Commit:
Implement PoseidonAnnotation & Combine logic

In order to let kelvind understand how to build the `Brach`/path
we need indeed to implement `Combine` and `Content` traits for
a custom type which is `PoseidonAnnotation`.

This now guarantees that kelvin trees will know how to build the
path and provide it correctly to us.

- Implemented `Combine` for `PoseidonAnnotation`. This contains the
logic for hashing an entire merkle tree level computing the bitflags
and then, returning the resulting value located on the second leaf.

- Implemented `Content` for `PoseidonAnnotation` since it's required
directly by kelvin.

Since it's needed to know when a leaf is empty, we need to
use `Option<T>` instead of `T` directly.

- Added lib exported structs and docs.

Closes #12

ZER0 commented on src/hashing_utils/poseidon_annotation.rs:
```diff
@@ -0,0 +1,92 @@
+        elements
+            .iter()
+            .zip(leaves.iter_mut())
+            .for_each(|(element, leave)| {
```
I'm totally fine to have it as is, but I'm wondering if what you're doing here it's actually mapping (and therefore a `map` method might be more clear on the code intention?)

CPerezz replied:
`for_each` seemed better to me since even I'm mapping, I'm not expecting to collect anything.

ZER0 commented on src/hashing_utils/poseidon_annotation.rs:
```diff
@@ -0,0 +1,92 @@
+            bytes[idx] = byte.unwrap();
+        }
+        Ok(PoseidonAnnotation(StorageScalar(
+            Scalar::from_bytes(&bytes).unwrap(),
```
Are we totally sure `unwrap` never fails (here and in line 62)…? Since we're already returning a `Result` maybe it might be worth to propagate / convert the error…?

CPerezz replied:
Yes since this is reading something that was already stored before. Therefore, if the scalar was transformed into bytes to be stored, it will be correct on the deserialization.

ZER0 commented on src/hashing_utils/scalar_storage.rs:
```diff
@@ -0,0 +1,52 @@
+///
+/// This Struct is the one that we will use inside of our SmartContract storage logic to
+/// encode/compress all of our Data Structures data into a single `Scalar`.
+pub struct StorageScalar(pub Scalar);
```
Is there a compelling reason why Scalar has to be public to everyone (and not just the crate)?

CPerezz replied:
Because this is what you're allowed to store inside Kelvin trees. You'll not be able to store `Scalar` directly.

Therefore I export it, so any library that deals with kelvin trees will need to implement the conversions from them custom types to this `StorageScalar`.

ZER0 commented on src/hashing_utils/scalar_storage.rs:
```diff
@@ -0,0 +1,52 @@
+    }
+}
+
+impl Into<Scalar> for StorageScalar {
```
Since `Into` usually should be avoided to be implemented directly, please add a comment about why here is needed.

CPerezz replied:
Done.

ZER0 commented on src/hashing_utils/scalar_storage.rs:
```diff
@@ -0,0 +1,52 @@
+    }
+    fn restore(source: &mut Source<H>) -> io::Result<Self> {
+        let mut bytes = [0u8; 32];
+        for (idx, byte) in source.bytes().enumerate() {
```
I assuming this is because you might have less than `32` bytes to iterate? Or you might have sparse index?

CPerezz replied:
I could have zipped them to make it easier, you're right, I'll refactor it!

ZER0 commented on src/lib.rs:
```diff
@@ -1,5 +1,38 @@
 #![deny(missing_docs)]
 
-/// Reference implementation for the Poseidon Sponge function
+///
```
Have you missing a comment here?

CPerezz replied:
Yup, I missed it, will be fixed by #19 which re-writes the lib.rs file properly.

ZER0 commented on src/merkle_lvl_hash/hash.rs:
```diff
@@ -0,0 +1,198 @@
+/// leaves and applies the poseidon hash, using the `hades252::ScalarStragegy` and
+/// computing the corresponding bitflags.
+#[allow(dead_code)]
+pub fn merkle_level_hash(leaves: &[Option<Scalar>]) -> Scalar {
```
I'm a bit confused by the names; it seems you're repeating yourself.
You're basically prefixing everything with `merkle_level_hash` that is more or less the name of the module, so why don't skip that? Instead of `merkle_lvl_hash::hash::merkle_level_hash` wouldn't be better `merkle_level::hash` and so on? 

Or I am missing something?

CPerezz replied:
Well, you're right on that.
The thing is that it's the name of the hashing technique. It's the same as saying `sponge_hash` in the `sponge` module.

I'm not sure on what to do. WDYT?

ZER0 commented on src/merkle_lvl_hash/hash.rs:
```diff
@@ -0,0 +1,198 @@
+#[allow(dead_code)]
+pub fn merkle_level_hash(leaves: &[Option<Scalar>]) -> Scalar {
+    let mut strategy = ScalarStrategy::new();
+    let mut accum = 0u64;
```
If you have an accumulator, does that means you're actually perform a folding instead of a for each?

CPerezz replied:
Ye, I will change it. My bad.

CPerezz replied:
Actually, it's a bit more trickier that what I expected. 
I'll leave it like this and open an issue for it wiht a low prio label.

ZER0 commented on src/merkle_proof/poseidon_branch.rs:
```diff
@@ -0,0 +1,134 @@
+            .unwrap()
+            .borrow()
+            .to_owned()
+            .into();
```
I assuming all the `unwrap`s are safe

krl replied:
prefer to use .expect instead of unwrap, and give a string that explains why it can never fail

CPerezz replied:
Which explanation will you suggest @krl ?

ZER0 commented on src/merkle_proof/poseidon_branch.rs:
```diff
@@ -0,0 +1,134 @@
+                    .for_each(|(idx, (src, dest))| {
+                        *dest = match src.annotation() {
+                            Some(borrow) => {
+                                let annotation: &Scalar = &(*borrow).borrow().borrow();
```
Is there no other way to make this line of code a bit more clear? :) I mean `&(*borrow).borrow().borrow()` it looks like tongue twister :D

krl replied:
i would recommend doing it in multiple lines (with type annotations) for clarity.

```
let b: &B = a.borrow();
let c: &C = c.borrow();
```

etc, also makes it easier to spot if there's one borrow too much (which is still legal code, just  more confusing)

CPerezz replied:
I like it Kris!

Done.

ZER0 commented on src/merkle_proof/poseidon_branch.rs:
```diff
@@ -0,0 +1,134 @@
+                                // A level with: [Some(val), None, None, None] should correspond to
+                                // Bitflags(1000). This means that we need to shift the first element
+                                // by `ARITY` and lately decrease the shift order by `idx`.
+                                level_bitflags += 1u64 << ((ARITY - 1) - idx);
```
If we're dealing with bitflag, shouldn't be `|=` for clarity (I'm pretty sure the compiler doesn't care much).
Also, the comment does not explain the `- 1`, it should be included.

CPerezz replied:
I'll address the `-1` nitpick.

Regarding the `|=` makes sense. If we deal with bits it's more accurate.

CPerezz replied:
Done.

ZER0 reviewed (changes requested):
Add some comments, check what you think it makes sense I trust you.
Also, since it involves Kelvin maybe a quick look from @krl might be useful as well.

Commit:
Addres #17 comments regarding StorageScalar

- Added `Into` impl info regarding why it's implemented.
- Support failures on reading from disk in `restore` impl
for `StorageScalar` propagating the errors.

Commit:
Address #17 comments for PoseidonBranch

- Added `Into` docs
- Added bitflags computation further explanations.
- Replaced `unwrap` for `expect` wit an unreachable error
message.
- Use | instead of + for bitflags accumulator.

CPerezz committed:
Merge branch 'master' into merkle_lvl_hash

krl commented on src/hashing_utils/poseidon_annotation.rs:
```diff
@@ -0,0 +1,92 @@
+    fn persist(&mut self, sink: &mut Sink<H>) -> io::Result<()> {
+        self.inner().0.to_bytes().persist(sink)
+    }
+    fn restore(source: &mut Source<H>) -> io::Result<Self> {
```
use read_exact, return InvalidData when from_bytes fails

CPerezz replied:
It should never fail right? SInce the bytes stored are already a `Scalar`.

For the read_exact it's ok. Unsure about the other thing.

krl replied:
from_bytes can fail according to the return type. Even if it never happens, it's still cleaner than using unwrap. IMO unwrap should never be used in production code.

If absolutely neccesary, use expect(), with a note explaining why

CPerezz replied:
Done with propper error handling. You were absolutely right!

krl commented on src/hashing_utils/scalar_storage.rs:
```diff
@@ -0,0 +1,58 @@
+    fn persist(&mut self, sink: &mut Sink<H>) -> io::Result<()> {
+        self.0.to_bytes().persist(sink)
+    }
+    fn restore(source: &mut Source<H>) -> io::Result<Self> {
```
use read_exact, return InvalidData...

CPerezz replied:
Same as avobe.

Commit:
Use read_exact and avoid panicking in Content impl

- As suggested by @krl, we should avoid panicking in
`Content` impls.

- Therefore we use read_exact and propagate the error
(if there is any).
- We generate an io::ErrorKind::InvealidData if the
generation of the `Scalar` from bytes fails.

This applies to `StorageScalar` and `PoseidonAnnotation`.

Commit:
Merge branch 'merkle_lvl_hash' of https://github.com/dusk-network/Poseidon252 into merkle_lvl_hash
